File: benchmarks/baselines/python/v2/fixture.py
"""Adapters between frozen JSON fixtures and the typed service boundary."""

from __future__ import annotations

import base64
import binascii
from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any, Never, cast

from v2.domain import (
    AlreadyExists,
    ApiVersion,
    Created,
    CreateJobRequest,
    CreateJobResult,
    InsertOutcome,
    Invalid,
    Job,
    PersistenceUnavailable,
    Priority,
)
from v2.service import create_job
from v2.store import DeterministicJobStore, RecordVersion, StoredJob

type JsonObject = dict[str, Any]


class FixtureError(ValueError):
    """A malformed or unsupported fixture boundary."""


@dataclass(frozen=True, slots=True)
class DecodedRequest:
    version: ApiVersion
    request: CreateJobRequest


def run_case(raw: object) -> JsonObject:
    fixture = _mapping(raw, "fixture")
    operation = _string(fixture, "operation")
    if operation == "create_job":
        return _run_create_case(fixture)
    if operation == "decode_stored_job":
        return _run_decode_stored_case(fixture)
    raise FixtureError(f"unsupported operation {operation!r}")


def _run_create_case(fixture: Mapping[str, object]) -> JsonObject:
    case_id = _string(fixture, "case_id")
    service_version = _record_version(_integer(fixture, "service_version"))
    initial_jobs = tuple(
        _parse_stored_job(value) for value in _list(fixture, "initial_jobs")
    )
    decoded, unknown_priority = _decode_request(
        service_version,
        _mapping(fixture.get("request"), "request"),
    )

    if unknown_priority is not None:
        actual: JsonObject = {
            "decode_error": {
                "code": "unknown_priority",
                "field": "priority",
                "value": unknown_priority,
            },
            "final_jobs": _encode_stored_jobs(initial_jobs),
            "store_calls": [],
        }
    else:
        outcome = _parse_store_outcome(fixture.get("store_outcome"))
        job_store = DeterministicJobStore(initial_jobs, outcome, service_version)
        result = create_job(decoded.request, job_store)
        actual = {
            "response": _encode_response(decoded.version, result),
            "final_jobs": _encode_stored_jobs(job_store.jobs),
            "store_calls": [
                {
                    "operation": "insert_if_absent",
                    "job": _encode_stored_job(call.job),
                }
                for call in job_store.calls
            ],
        }
    return {
        "result_format": 1,
        "case_id": case_id,
        "operation": "create_job",
        "actual": actual,
    }


def _run_decode_stored_case(fixture: Mapping[str, object]) -> JsonObject:
    stored = _parse_stored_job(_mapping(fixture.get("stored_job"), "stored_job"))
    return {
        "result_format": 1,
        "case_id": _string(fixture, "case_id"),
        "operation": "decode_stored_job",
        "actual": {"decoded_job": _encode_job_v2(stored.adapt_to_v2())},
    }


def _decode_request(
    service_version: RecordVersion,
    raw: Mapping[str, object],
) -> tuple[DecodedRequest, str | None]:
    try:
        version = ApiVersion(_integer(raw, "api_version"))
    except ValueError as error:
        raise FixtureError("unsupported request API version") from error
    if service_version is RecordVersion.V1 and version is not ApiVersion.V1:
        raise FixtureError("service version 1 accepts only API version 1")

    unknown_priority: str | None = None
    if version is ApiVersion.V1:
        priority: Priority | None = Priority.NORMAL
    elif "priority" not in raw:
        priority = None
    else:
        value = _string(raw, "priority")
        try:
            priority = Priority(value)
        except ValueError:
            priority = None
            unknown_priority = value

    request = CreateJobRequest(
        job_id=_string(raw, "job_id"),
        task=_string(raw, "task"),
        payload=_decode_base64(_string(raw, "payload_base64"), "payload"),
        priority=priority,
    )
    return DecodedRequest(version, request), unknown_priority


def _parse_stored_job(raw: object) -> StoredJob:
    value = _mapping(raw, "stored job")
    version = _record_version(_integer(value, "record_version"))
    priority: Priority | None = None
    if version is RecordVersion.V2:
        if "priority" not in value:
            raise FixtureError("version-two stored job is missing priority")
        raw_priority = _string(value, "priority")
        try:
            priority = Priority(raw_priority)
        except ValueError as error:
            raise FixtureError(
                f"version-two stored job has unknown priority {raw_priority!r}"
            ) from error
    return StoredJob(
        record_version=version,
        job_id=_string(value, "job_id"),
        task=_string(value, "task"),
        payload=_decode_base64(
            _string(value, "payload_base64"),
            "stored payload",
        ),
        priority=priority,
    )


def _parse_store_outcome(value: object) -> InsertOutcome:
    if value is None:
        return InsertOutcome.UNAVAILABLE_BEFORE_COMMIT
    if not isinstance(value, str):
        raise FixtureError("store_outcome must be a string")
    try:
        return InsertOutcome(value)
    except ValueError as error:
        raise FixtureError(f"unsupported store outcome {value!r}") from error


def _record_version(value: int) -> RecordVersion:
    try:
        return RecordVersion(value)
    except ValueError as error:
        raise FixtureError(f"unsupported service or record version {value}") from error


def _encode_response(
    version: ApiVersion,
    result: CreateJobResult,
) -> JsonObject:
    match result:
        case Created(job):
            encoded: JsonObject = {
                "kind": "created",
                "job": (
                    {
                        "job_id": job.job_id,
                        "task": job.task,
                        "payload_base64": _encode_base64(job.payload),
                    }
                    if version is ApiVersion.V1
                    else {
                        "job_id": job.job_id,
                        "task": job.task,
                        "payload_base64": _encode_base64(job.payload),
                        "priority": job.priority.value,
                    }
                ),
            }
        case Invalid(issues):
            encoded = {
                "kind": "invalid",
                "issues": [
                    {"field": issue.field.value, "reason": issue.reason.value}
                    for issue in issues
                ],
            }
        case AlreadyExists(job_id):
            encoded = {"kind": "already_exists", "job_id": job_id}
        case PersistenceUnavailable():
            encoded = {"kind": "persistence_unavailable"}
        case _:
            _assert_never(result)
    return {"api_version": int(version), "result": encoded}


def _encode_job_v2(job: Job) -> JsonObject:
    return {
        "record_version": 2,
        "job_id": job.job_id,
        "task": job.task,
        "payload_base64": _encode_base64(job.payload),
        "priority": job.priority.value,
    }


def _encode_stored_jobs(jobs: tuple[StoredJob, ...]) -> list[JsonObject]:
    return [_encode_stored_job(job) for job in jobs]


def _encode_stored_job(job: StoredJob) -> JsonObject:
    result: JsonObject = {
        "record_version": int(job.record_version),
        "job_id": job.job_id,
        "task": job.task,
        "payload_base64": _encode_base64(job.payload),
    }
    if job.record_version is RecordVersion.V2:
        if job.priority is None:
            msg = "version-two stored job must have priority"
            raise AssertionError(msg)
        result["priority"] = job.priority.value
    return result


def _encode_base64(value: bytes) -> str:
    return base64.b64encode(value).decode("ascii")


def _decode_base64(value: str, label: str) -> bytes:
    try:
        return base64.b64decode(value, validate=True)
    except (binascii.Error, ValueError) as error:
        raise FixtureError(f"invalid {label} Base64") from error


def _mapping(value: object, label: str) -> Mapping[str, object]:
    if not isinstance(value, Mapping):
        raise FixtureError(f"{label} must be an object")
    return cast("Mapping[str, object]", value)


def _list(value: Mapping[str, object], key: str) -> list[object]:
    result = value.get(key)
    if not isinstance(result, list):
        raise FixtureError(f"{key} must be an array")
    return cast("list[object]", result)


def _string(value: Mapping[str, object], key: str) -> str:
    result = value.get(key)
    if not isinstance(result, str):
        raise FixtureError(f"{key} must be a string")
    return result


def _integer(value: Mapping[str, object], key: str) -> int:
    result = value.get(key)
    if not isinstance(result, int) or isinstance(result, bool):
        raise FixtureError(f"{key} must be an integer")
    return result


def _assert_never(value: Never) -> Never:
    raise AssertionError(f"unhandled create-job result {value!r}")

File: benchmarks/baselines/typescript/v2/fixture.ts
/** Frozen JSON fixture adapters for the typed service boundary. */

import {
  ApiVersion,
  InsertOutcome,
  Priority,
  parsePriority,
  type CreateJobRequest,
  type CreateJobResult,
  type Job,
} from "./domain.js";
import { createJob } from "./service.js";
import {
  DeterministicJobStore,
  RecordVersion,
  adaptStoredJobToV2,
  type RecordVersion as RecordVersionType,
  type StoredJob,
} from "./store.js";

export type JsonObject = Record<string, unknown>;

export class FixtureError extends Error {}

export function runCase(raw: unknown): JsonObject {
  const fixture = object(raw, "fixture");
  const operation = string(fixture, "operation");
  if (operation === "create_job") {
    return runCreateCase(fixture);
  }
  if (operation === "decode_stored_job") {
    return runDecodeStoredCase(fixture);
  }
  throw new FixtureError(`unsupported operation ${JSON.stringify(operation)}`);
}

function runCreateCase(fixture: JsonObject): JsonObject {
  const caseId = string(fixture, "case_id");
  const serviceVersion = recordVersion(integer(fixture, "service_version"));
  const initialJobs = array(fixture, "initial_jobs").map(parseStoredJob);
  const { decoded, unknownPriority } = decodeRequest(
    serviceVersion,
    object(fixture.request, "request"),
  );

  let actual: JsonObject;
  if (unknownPriority !== undefined) {
    actual = {
      decode_error: {
        code: "unknown_priority",
        field: "priority",
        value: unknownPriority,
      },
      final_jobs: initialJobs.map(encodeStoredJob),
      store_calls: [],
    };
  } else {
    const outcome = parseStoreOutcome(fixture.store_outcome);
    const store = new DeterministicJobStore(
      initialJobs,
      outcome,
      serviceVersion,
    );
    const result = createJob(decoded.request, store);
    actual = {
      response: encodeResponse(decoded.version, result),
      final_jobs: store.jobs.map(encodeStoredJob),
      store_calls: store.calls.map((call) => ({
        operation: "insert_if_absent",
        job: encodeStoredJob(call.job),
      })),
    };
  }
  return {
    result_format: 1,
    case_id: caseId,
    operation: "create_job",
    actual,
  };
}

function runDecodeStoredCase(fixture: JsonObject): JsonObject {
  const stored = parseStoredJob(object(fixture.stored_job, "stored_job"));
  return {
    result_format: 1,
    case_id: string(fixture, "case_id"),
    operation: "decode_stored_job",
    actual: { decoded_job: encodeJobV2(adaptStoredJobToV2(stored)) },
  };
}

function decodeRequest(
  serviceVersion: RecordVersionType,
  raw: JsonObject,
): Readonly<{
  decoded: Readonly<{ version: ApiVersion; request: CreateJobRequest }>;
  unknownPriority: string | undefined;
}> {
  const rawVersion = integer(raw, "api_version");
  if (rawVersion !== ApiVersion.v1 && rawVersion !== ApiVersion.v2) {
    throw new FixtureError(`unsupported request API version ${rawVersion}`);
  }
  const version: ApiVersion = rawVersion;
  if (serviceVersion === RecordVersion.v1 && version !== ApiVersion.v1) {
    throw new FixtureError("service version 1 accepts only API version 1");
  }

  let priority: Priority | undefined;
  let unknownPriority: string | undefined;
  if (version === ApiVersion.v1) {
    priority = Priority.normal;
  } else if ("priority" in raw) {
    const value = string(raw, "priority");
    priority = parsePriority(value);
    if (priority === undefined) {
      unknownPriority = value;
    }
  }

  return {
    decoded: {
      version,
      request: {
        jobId: string(raw, "job_id"),
        task: string(raw, "task"),
        payload: decodeBase64(string(raw, "payload_base64"), "payload"),
        priority,
      },
    },
    unknownPriority,
  };
}

function parseStoredJob(raw: unknown): StoredJob {
  const value = object(raw, "stored job");
  const version = recordVersion(integer(value, "record_version"));
  let priority: Priority | undefined;
  if (version === RecordVersion.v2) {
    if (!("priority" in value)) {
      throw new FixtureError("version-two stored job is missing priority");
    }
    const rawPriority = string(value, "priority");
    priority = parsePriority(rawPriority);
    if (priority === undefined) {
      throw new FixtureError(
        `version-two stored job has unknown priority ${JSON.stringify(rawPriority)}`,
      );
    }
  }
  return {
    recordVersion: version,
    jobId: string(value, "job_id"),
    task: string(value, "task"),
    payload: decodeBase64(string(value, "payload_base64"), "stored payload"),
    priority,
  };
}

function parseStoreOutcome(value: unknown): InsertOutcome {
  if (value === undefined) {
    return InsertOutcome.unavailableBeforeCommit;
  }
  if (
    value === InsertOutcome.inserted ||
    value === InsertOutcome.duplicate ||
    value === InsertOutcome.unavailableBeforeCommit
  ) {
    return value;
  }
  throw new FixtureError(`unsupported store outcome ${JSON.stringify(value)}`);
}

function recordVersion(value: number): RecordVersionType {
  if (value === RecordVersion.v1 || value === RecordVersion.v2) {
    return value;
  }
  throw new FixtureError(`unsupported service or record version ${value}`);
}

function encodeResponse(
  version: ApiVersion,
  result: CreateJobResult,
): JsonObject {
  let encoded: JsonObject;
  switch (result.kind) {
    case "created":
      encoded = {
        kind: "created",
        job:
          version === ApiVersion.v1
            ? {
                job_id: result.job.jobId,
                task: result.job.task,
                payload_base64: encodeBase64(result.job.payload),
              }
            : {
                job_id: result.job.jobId,
                task: result.job.task,
                payload_base64: encodeBase64(result.job.payload),
                priority: result.job.priority,
              },
      };
      break;
    case "invalid":
      encoded = {
        kind: "invalid",
        issues: result.issues.map((issue) => ({ ...issue })),
      };
      break;
    case "already_exists":
      encoded = { kind: "already_exists", job_id: result.jobId };
      break;
    case "persistence_unavailable":
      encoded = { kind: "persistence_unavailable" };
      break;
    default:
      return assertNever(result);
  }
  return { api_version: version, result: encoded };
}

function encodeJobV2(job: Job): JsonObject {
  return {
    record_version: 2,
    job_id: job.jobId,
    task: job.task,
    payload_base64: encodeBase64(job.payload),
    priority: job.priority,
  };
}

function encodeStoredJob(job: StoredJob): JsonObject {
  const result: JsonObject = {
    record_version: job.recordVersion,
    job_id: job.jobId,
    task: job.task,
    payload_base64: encodeBase64(job.payload),
  };
  if (job.recordVersion === RecordVersion.v2) {
    if (job.priority === undefined) {
      throw new Error("version-two stored job must have priority");
    }
    result.priority = job.priority;
  }
  return result;
}

function decodeBase64(value: string, label: string): Uint8Array {
  if (
    value.length % 4 !== 0 ||
    !/^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/u.test(
      value,
    )
  ) {
    throw new FixtureError(`invalid ${label} Base64`);
  }
  return Uint8Array.from(Buffer.from(value, "base64"));
}

function encodeBase64(value: Uint8Array): string {
  return Buffer.from(value).toString("base64");
}

function object(value: unknown, label: string): JsonObject {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new FixtureError(`${label} must be an object`);
  }
  return value as JsonObject;
}

function array(value: JsonObject, key: string): readonly unknown[] {
  const result = value[key];
  if (!Array.isArray(result)) {
    throw new FixtureError(`${key} must be an array`);
  }
  return result;
}

function string(value: JsonObject, key: string): string {
  const result = value[key];
  if (typeof result !== "string") {
    throw new FixtureError(`${key} must be a string`);
  }
  return result;
}

function integer(value: JsonObject, key: string): number {
  const result = value[key];
  if (!Number.isInteger(result)) {
    throw new FixtureError(`${key} must be an integer`);
  }
  return result as number;
}

function assertNever(value: never): never {
  throw new Error(`unhandled create-job result ${JSON.stringify(value)}`);
}
